src/librustc_target/abi/call/riscv.rs

```diff
@@ -1,49 +1,341 @@
 // Reference: RISC-V ELF psABI specification
 // https://github.com/riscv/riscv-elf-psabi-doc
+//
+// Reference: Clang RISC-V ELF psABI lowering code
+// https://github.com/llvm/llvm-project/blob/8e780252a7284be45cf1ba224cabd884847e8e92/clang/lib/CodeGen/TargetInfo.cpp#L9311-L9773
 
-use crate::abi::call::{ArgAbi, FnAbi};
+use crate::abi::call::{ArgAbi, ArgAttribute, FnAbi, PassMode, Reg, Uniform};
+use crate::abi::{self, Abi, FieldPlacement, LayoutOf, TyLayout, TyLayoutMethods};
+use crate::spec::HasTargetSpec;
+
+#[derive(Copy, Clone)]
+enum FieldKind {
+    Float,
+    Integer,
+    Unknown,
+}
```

Comment:
I would suggest using `Option<RegKind>` instead.

Reply:
And if you need to use `CastTarget`, might as well keep a full `Option<Reg>`.

Reply:
The code still need to know what type the register is so it can track the counts of remaining registers.

Reply:
`RegKind` already has that information.

Reply:
Maybe, but it would be shorter than trying to re-extract the information from the `Reg` below.